crates/wasm-encoder/src/reencode.rs

```diff
@@ -0,0 +1,1308 @@
+#![allow(missing_docs)] // FIXME
+
+#[derive(Debug)]
+pub enum Error {
+    /// There was a type reference that was canonicalized and no longer
+    /// references an index into a module's types space, so we cannot encode it
+    /// into a Wasm binary again.
+    CanonicalizedHeapTypeReference,
+    /// The const expression is invalid: not actually constant or something like
+    /// that.
+    InvalidConstExpr,
+    /// There was an error when parsing.
+    ParseError(wasmparser::BinaryReaderError),
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::ParseError(_e) => {
+                write!(fmt, "There was an error when parsing")
+            }
+            Self::InvalidConstExpr => write!(fmt, "The const expression was invalid"),
+            Self::CanonicalizedHeapTypeReference => write!(
+                fmt,
+                "There was a canonicalized heap type reference without type index information"
+            ),
+        }
+    }
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Self::ParseError(e) => Some(e),
+            Self::InvalidConstExpr | Self::CanonicalizedHeapTypeReference => None,
+        }
+    }
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub trait WasmParserToWasmEncoder {
```

Comment:
I like the name "reencode" for this module so perhaps this trait could be `Reencode` as well?

Reply:
Done